drv/fpga-devices/src/ecp5_spi_mux_pca9538.rs

```diff
@@ -0,0 +1,297 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use crate::ecp5::{Command, Ecp5Driver};
+use crate::FpgaUserDesign;
+use drv_fpga_api::FpgaError;
+use drv_i2c_api::ResponseCode;
+use drv_i2c_devices::pca9538;
+use drv_spi_api::{self as spi_api, SpiDevice, SpiError};
+use drv_stm32xx_sys_api::{self as sys_api, GpioError, Sys};
+
+/// This module implements an ECP5 driver which exposes two physical devices,
+/// which share a single SPI bus using a mux and are controlled through a shared
+/// PCA9538 GPIO expander.
+
+/// Impl Error type, with conversion from `GpioError`, `SpiError` and
+/// `ResponseCode`.
+#[derive(Copy, Clone, Debug)]
+pub enum Error {
+    GpioError(GpioError),
+    SpiError(SpiError),
+    I2cError(ResponseCode),
+}
+
+impl From<GpioError> for Error {
+    fn from(e: GpioError) -> Self {
+        Self::GpioError(e)
+    }
+}
+
+impl From<SpiError> for Error {
+    fn from(e: SpiError) -> Self {
+        Self::SpiError(e)
+    }
+}
+
+impl From<ResponseCode> for Error {
+    fn from(e: ResponseCode) -> Self {
+        Self::I2cError(e)
+    }
+}
+
+impl From<Error> for u8 {
+    fn from(e: Error) -> Self {
+        match e {
+            Error::GpioError(e) => match e {
+                GpioError::BadArg => 2,
+            },
+            Error::SpiError(e) => match e {
+                SpiError::BadTransferSize => 3,
+                SpiError::ServerRestarted => 4,
+                SpiError::NothingToRelease => 5,
+                SpiError::BadDevice => 6,
+                SpiError::DataOverrun => 7,
+            },
+            Error::I2cError(e) => 8 + (e as u8),
+        }
+    }
+}
+
+impl From<Error> for FpgaError {
+    fn from(e: Error) -> Self {
+        FpgaError::ImplError(u8::from(e))
+    }
+}
+
+pub struct Pins {
+    pub done: pca9538::PinSet,
+    pub init_n: pca9538::PinSet,
+    pub program_n: pca9538::PinSet,
+    pub user_design_reset_n: pca9538::PinSet,
+}
+
+pub struct Driver {
+    pub sys: Sys,
+    pub configuration_port: SpiDevice,
+    pub user_design: SpiDevice,
+    pub spi_mux_select: sys_api::PinSet,
+    pub instance_selected: core::cell::Cell<usize>,
+    pub gpio: pca9538::Pca9538,
+    pub user_design_reset_duration: u64,
+}
+
+impl Driver {
+    pub fn select_instance(&self, instance: usize) -> Result<(), GpioError> {
+        if self.instance_selected.get() != instance {
+            self.instance_selected.set(instance);
+            self.sys.gpio_set_to(
```

Comment:
I like to store `Cell<Option<usize>>` for caches like this one.

That way, if `gpio_set_to` returns an error, we'd store `None` in the cache, representing an unknown state (since we don't know if the GPIO task crashed before or after setting the pin).

Reply:
I updated the function with the suggestion, but let me know if there is a more concise way to do this.